### ISSUE TITLE
Compilation fails at geant4 example with multi-threaded local install of geant4

### ISSUE THREAD

**Issue by php1ic** [mctools/ncrystal#35]
I'm on a fully updated Debian 9 and have geant4 installed using the following
```
cmake -DCMAKE_INSTALL_PREFIX=/usr/local/share/geant4/10.4.0/install/debug -DGEANT4_INSTALL_DATADIR=/usr/local/share/geant4/data -DGEANT4_BUILD_MULTITHREADED=ON -DCMAKE_BUILD_TYPE=RelWithDebInfo -DGEANT4_USE_QT=ON -DGEANT4_USE_GDML=ON -DGEANT4_USE_OPENGL_X11=ON -DGEANT4_USE_RAYTRACER_X11=ON /usr/local/share/geant4/10.4.0/geant4.10.04
```

I do the following from inside a fresh clone of ncrystal
```
$ mkdir build
$ cd build
$ cmake ../
-- The CXX compiler identification is GNU 6.3.0
-- The C compiler identification is GNU 6.3.0
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Found PythonInterp: /usr/bin/python (found version "2.7.13") 
###################################################
## NCrystal configuration summary:               ##
##   NCrystal library and headers       : yes    ##
##   NCrystal python module and scripts : yes    ##
##   G4NCrystal library and headers     : yes    ##
##   Enable examples for C and C++      : yes    ##
##   Install shipped data files         : yes    ##
##   Enable .nxs/.laz/.lau support      : yes    ##
###################################################
-- Configuring done
-- Generating done
-- Build files have been written to: /home/php1ic/repos/ncrystal/build
```

but the build fails at the linking stage of the geant4 example
```
$ make VERBOSE=1
.
.
.
[ 95%] Linking CXX executable ncrystal_example_g4sim
/usr/bin/cmake -E cmake_link_script CMakeFiles/ncrystal_example_g4sim.dir/link.txt --verbose=1
/usr/bin/c++   -O3 -DNDEBUG   CMakeFiles/ncrystal_example_g4sim.dir/examples/ncrystal_example_g4sim.cc.o  -o ncrystal_example_g4sim -Wl,-rpath,/home/php1ic/repos/ncrystal/build:/usr/local/share/geant4/10.4.0/install/debug/lib: -rdynamic libG4NCrystal.so libNCrystal.so -lm /usr/local/share/geant4/10.4.0/install/debug/lib/libG4Tree.so /usr/local/share/geant4/10.4.0/install/debug/lib/libG4GMocren.so /usr/local/share/geant4/10.4.0/install/debug/lib/libG4FR.so /usr/local/share/geant4/10.4.0/install/debug/lib/libG4visHepRep.so /usr/local/share/geant4/10.4.0/install/debug/lib/libG4RayTracer.so -lSM -lICE -lX11 -lXext /usr/local/share/geant4/10.4.0/install/debug/lib/libG4VRML.so /usr/local/share/geant4/10.4.0/install/debug/lib/libG4vis_management.so /usr/local/share/geant4/10.4.0/install/debug/lib/libG4modeling.so /usr/local/share/geant4/10.4.0/install/debug/lib/libG4interfaces.so /usr/lib/x86_64-linux-gnu/libQt5PrintSupport.so.5.7.1 /usr/lib/x86_64-linux-gnu/libQt5Widgets.so.5.7.1 /usr/lib/x86_64-linux-gnu/libQt5Gui.so.5.7.1 /usr/lib/x86_64-linux-gnu/libQt5Core.so.5.7.1 /usr/local/share/geant4/10.4.0/install/debug/lib/libG4persistency.so -lxerces-c /usr/local/share/geant4/10.4.0/install/debug/lib/libG4error_propagation.so /usr/local/share/geant4/10.4.0/install/debug/lib/libG4readout.so /usr/local/share/geant4/10.4.0/install/debug/lib/libG4physicslists.so /usr/local/share/geant4/10.4.0/install/debug/lib/libG4run.so /usr/local/share/geant4/10.4.0/install/debug/lib/libG4event.so /usr/local/share/geant4/10.4.0/install/debug/lib/libG4tracking.so /usr/local/share/geant4/10.4.0/install/debug/lib/libG4parmodels.so /usr/local/share/geant4/10.4.0/install/debug/lib/libG4processes.so /usr/local/share/geant4/10.4.0/install/debug/lib/libG4analysis.so -lexpat /usr/local/share/geant4/10.4.0/install/debug/lib/libG4digits_hits.so /usr/local/share/geant4/10.4.0/install/debug/lib/libG4track.so /usr/local/share/geant4/10.4.0/install/debug/lib/libG4particles.so /usr/local/share/geant4/10.4.0/install/debug/lib/libG4geometry.so /usr/local/share/geant4/10.4.0/install/debug/lib/libG4materials.so /usr/local/share/geant4/10.4.0/install/debug/lib/libG4graphics_reps.so /usr/local/share/geant4/10.4.0/install/debug/lib/libG4intercoms.so /usr/local/share/geant4/10.4.0/install/debug/lib/libG4global.so /usr/local/share/geant4/10.4.0/install/debug/lib/libG4clhep.so /usr/local/share/geant4/10.4.0/install/debug/lib/libG4zlib.so 
libG4NCrystal.so: undefined reference to `G4cout'
collect2: error: ld returned 1 exit status
CMakeFiles/ncrystal_example_g4sim.dir/build.make:135: recipe for target 'ncrystal_example_g4sim' failed
make[2]: *** [ncrystal_example_g4sim] Error 1
make[2]: Leaving directory '/home/php1ic/repos/ncrystal/build'
CMakeFiles/Makefile2:105: recipe for target 'CMakeFiles/ncrystal_example_g4sim.dir/all' failed
make[1]: *** [CMakeFiles/ncrystal_example_g4sim.dir/all] Error 2
make[1]: Leaving directory '/home/php1ic/repos/ncrystal/build'
Makefile:127: recipe for target 'all' failed
make: *** [all] Error 2
```

Can ncrystal only be used with non mutlithreaded installs of geant4?

**Comment by tkittel** [comment]
Hi @php1ic . Thanks for reporting this issue. Yes, admittedly NCrystal is currently not believed to work with Multi-threaded Geant4 (simply due to not time in the day, not because we don't think it is important). I can't give a definite timeline for when this will be fixed (cf. #18) , but I really hope to get around to it "sometime next year".

We are ourselves using multi-processing in our Geant4 jobs, which explains why we haven't paid enough attention to this, sorry.

**Comment by tkittel** [comment]
I'll keep this issue open, to remind us.

**Comment by tkittel** [comment]
I should also mention that there is a small possibility that your build error is unrelated and it is actually OK to build and use NCrystal against an installation of Geant4 with MT enabled, as long as one doesn't actually use the MT run manager to launch jobs.

However, this is not something I have investigated further.

**Comment by php1ic** [comment]
Hi @tkittel thanks for the quick response!

I can get things to build, then install, setup and run the geant4 example (using my MT version of geant4) with the following patch to the top level CMakeLists.txt file:
```
diff --git a/CMakeLists.txt b/CMakeLists.txt
index 19d8e81..0cee28f 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -132,6 +132,7 @@ else()
 endif()
 
 if (Geant4_FOUND)
+  include(${Geant4_USE_FILE})
   add_library(G4NCrystal SHARED ${SRCS_NCG4})
   target_include_directories(G4NCrystal PUBLIC "${CMAKE_CURRENT_SOURCE_DIR}/ncrystal_geant4/include")
   target_link_libraries(G4NCrystal NCrystal ${Geant4_LIBRARIES})
```

I'm not a cmake expert, but I think this additional line tells the ncrystal build step to use the same flags as those used to compile the referenced version of geant4, so pulls in the MT flag making the built .so library reference the correct version of G4cout (which was causing my initial error)

https://geant4.kek.jp/Reference/10.05.p01/G4ios_8hh_source.html

Not sure how safe this change is, it may cause issues in other places.

I currently make heavy use of the MT run manager, so look forward to "sometime next year" :wink:

**Comment by tkittel** [comment]
Ok, thanks for debugging further. So I think this makes sense: We don't compile with the correct state of the G4MULTITHREADED flag, so see a different G4cout that the G4 libraries saw when they were compiled.

However, including Geant4_USE_FILE has global effects, also affecting (if I understand correctly) other build targets, rather than only affecting the Geant4-related targets. If you can spare the time, can you try if it works if you instead put:
```
   target_compile_definitions(G4NCrystal PUBLIC ${Geant4_DEFINITIONS})
```
It might also be that you have to change the (now obsolete) line a bit further down from:
```
   target_compile_options(G4NCrystal PUBLIC "-std=c++11")
```
to
```
   target_compile_options(G4NCrystal PUBLIC ${Geant4_CXX_FLAGS})
```
but I am less sure about that one. I have to run now, but if you can confirm that this at least makes it work for you when not using G4MTRunManager, then I can put that code into the develop branch asap tomorrow, so you don't have to apply local hacks to make it compile for you.

That you actually also would like to run in MT mode eventually is duly noted of course :-)

**Comment by php1ic** [comment]
If I do both changes, compilation fails with:
```
[ 88%] Building CXX object CMakeFiles/G4NCrystal.dir/ncrystal_geant4/src/G4NCInstall.cc.o
/usr/bin/c++   -DG4LIB_BUILD_DLL -DG4MULTITHREADED -DG4NCrystal_EXPORTS -DG4UI_USE -DG4VERBOSE -DG4VIS_USE -DG4_STORE_TRAJECTORY -DQT_CORE_LIB -DQT_GUI_LIB -DQT_NO_DEBUG -DQT_PRINTSUPPORT_LIB -DQT_WIDGETS_LIB -I/home/icullen/repos/ncrystal/ncrystal_geant4/include -isystem /usr/local/share/geant4/10.5.1/install/debug/include/Geant4 -I/home/icullen/repos/ncrystal/ncrystal_core/include -isystem /usr/include/x86_64-linux-gnu/qt5 -isystem /usr/include/x86_64-linux-gnu/qt5/QtWidgets -isystem /usr/include/x86_64-linux-gnu/qt5/QtGui -isystem /usr/include/x86_64-linux-gnu/qt5/QtCore -isystem /usr/lib/x86_64-linux-gnu/qt5/mkspecs/linux-g++-64 -isystem /usr/include/x86_64-linux-gnu/qt5/QtPrintSupport  -O3 -DNDEBUG -fPIC   "-W -Wall -pedantic -Wno-non-virtual-dtor -Wno-long-long -Wwrite-strings -Wpointer-arith -Woverloaded-virtual -Wno-variadic-macros -Wshadow -pipe -DG4USE_STD11 -pthread -ftls-model=initial-exec -std=c++11" -fPIC -std=c++11 -o CMakeFiles/G4NCrystal.dir/ncrystal_geant4/src/G4NCInstall.cc.o -c /home/icullen/repos/ncrystal/ncrystal_geant4/src/G4NCInstall.cc
c++: error: unrecognized command line option '-W -Wall -pedantic -Wno-non-virtual-dtor -Wno-long-long -Wwrite-strings -Wpointer-arith -Woverloaded-virtual -Wno-variadic-macros -Wshadow -pipe -DG4USE_STD11 -pthread -ftls-model=initial-exec -std=c++11'
CMakeFiles/G4NCrystal.dir/build.make:62: recipe for target 'CMakeFiles/G4NCrystal.dir/ncrystal_geant4/src/G4NCInstall.cc.o' failed
make[2]: *** [CMakeFiles/G4NCrystal.dir/ncrystal_geant4/src/G4NCInstall.cc.o] Error 1
make[2]: Leaving directory '/home/icullen/repos/build/ncrystal'
CMakeFiles/Makefile2:142: recipe for target 'CMakeFiles/G4NCrystal.dir/all' failed
make[1]: *** [CMakeFiles/G4NCrystal.dir/all] Error 2
make[1]: Leaving directory '/home/icullen/repos/build/ncrystal'
Makefile:127: recipe for target 'all' failed
make: *** [all] Error 2
```
Looks like the different compiler flags are being passed as a single string argument for some reason?

Applying just the `target_compile_definition` change things compile and work.

**Comment by tkittel** [comment]
So on my Ubuntu I actually did not link successfully without also the contents from Geant4_CXX_FLAGS, since it contains a `-DG4USE_STD11` which was absent from Geant4_DEFINITIONS. Seems a bit buggy to be honest.

So here is the full change that worked for me (currently added to the develop branch, but will eventually make it into master and a new release):

```diff
diff --git a/CMakeLists.txt b/CMakeLists.txt
index 19d8e81..c1f91f2 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -13,7 +13,7 @@
 #                                                                                #
 #  $> make install                                                               #
 #                                                                                #
-# Written 2016-2017 by T. Kittelmann.                                            #
+# Written 2016-2019 by T. Kittelmann.                                            #
 #                                                                                #
 ##################################################################################
 
@@ -132,15 +132,19 @@ else()
 endif()
 
 if (Geant4_FOUND)
+  #Build G4NCrystal library:
   add_library(G4NCrystal SHARED ${SRCS_NCG4})
   target_include_directories(G4NCrystal PUBLIC "${CMAKE_CURRENT_SOURCE_DIR}/ncrystal_geant4/include")
-  target_link_libraries(G4NCrystal NCrystal ${Geant4_LIBRARIES})
-  target_include_directories(G4NCrystal SYSTEM PUBLIC ${Geant4_INCLUDE_DIRS})
-  #Recent G4 versions all requires C++11 (we enable it like this for simplicity,
-  #since there is not yet a standard way to enable C++11 with CMake - until CMake 3.1):
-  target_compile_options(G4NCrystal PUBLIC "-std=c++11")
   install(TARGETS G4NCrystal ${INSTDEST})
   install(FILES ${HDRS_NCG4} DESTINATION include/G4NCrystal)
+  #Transfer G4 flags uncovered by find_package(Geant4) call:
+  target_compile_definitions(G4NCrystal PUBLIC ${Geant4_DEFINITIONS})
+  target_link_libraries(G4NCrystal NCrystal ${Geant4_LIBRARIES})
+  target_include_directories(G4NCrystal SYSTEM PUBLIC ${Geant4_INCLUDE_DIRS})
+  set(Geant4_CXX_FLAGS_aslist ${Geant4_CXX_FLAGS})
+  separate_arguments(Geant4_CXX_FLAGS_aslist)
+  target_compile_options(G4NCrystal PUBLIC ${Geant4_CXX_FLAGS_aslist})
+  #examples if needed:
   if (BUILD_EXAMPLES AND EXAMPLES_NCG4)
     foreach(ex ${EXAMPLES_NCG4})
       get_filename_component(exbn "${ex}" NAME_WE)
```

**Comment by tkittel** [comment]
Keeping this issue open until the fix propagates into an actual release.

I opened #36 to separately track the issue of actually supporting multithreaded Geant4 runs.

**Comment by tkittel** [comment]
Fix is now in release 2.0.0, closing here.
